File: libs/ktheme-swift/Sources/Ktheme/Parser/ThemeParser.swift
// libs/ktheme-swift/Sources/Ktheme/Parser/ThemeParser.swift
//
// Tolerant Theme JSON decode/encode. Lenient enough that older preset
// files (missing `adaptation`, missing `shimmer`) parse cleanly.

import Foundation

public enum ThemeParser {

    public static let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .useDefaultKeys
        return d
    }()

    public static let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.outputFormatting = [.prettyPrinted, .sortedKeys]
        e.keyEncodingStrategy = .useDefaultKeys
        return e
    }()

    public static func decode(_ json: String) throws -> Theme {
        try decoder.decode(Theme.self, from: Data(json.utf8))
    }
    public static func decode(_ data: Data) throws -> Theme {
        try decoder.decode(Theme.self, from: data)
    }
    public static func encode(_ theme: Theme) throws -> String {
        let data = try encoder.encode(theme)
        return String(data: data, encoding: .utf8) ?? "{}"
    }
}

File: libs/ktheme-core/src/main/kotlin/io/ktheme/parser/ThemeParser.kt
// libs/ktheme-core/src/main/kotlin/io/ktheme/parser/ThemeParser.kt
//
// Single entry point for reading + writing Theme JSON. Kept lenient so
// older preset files (missing `adaptation`, missing `shimmer`, etc.)
// parse without exceptions.

package io.ktheme.parser

import io.ktheme.model.Theme
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

public object ThemeParser {

    /**
     * Strict-but-tolerant JSON config:
     *  - `ignoreUnknownKeys` — preset files often add experimental fields.
     *  - `coerceInputValues` — null primitives fall back to defaults.
     *  - `prettyPrint` for round-tripping; consumers who want compact
     *     can call [Json] themselves.
     */
    public val Json: Json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
        encodeDefaults = false
        prettyPrint = true
        prettyPrintIndent = "  "
    }

    /** Decode a single theme from a JSON string. */
    public fun decode(jsonString: String): Theme =
        Json.decodeFromString(Theme.serializer(), jsonString)

    /** Encode a single theme as a JSON string (pretty-printed). */
    public fun encode(theme: Theme): String =
        Json.encodeToString(Theme.serializer(), theme)

    /** Decode a list of themes (e.g. an exported pack). */
    public fun decodeAll(jsonString: String): List<Theme> =
        Json.decodeFromString(jsonString)
}
